at.1.in
.\"     %A%
.\"
.Dd August 12, 2025
.Dt AT 1
.Os MASTODON
.Sh NAME
.Nm at
.Nd execute commands at a specified time
.Sh SYNOPSIS
.Nm
.Op Fl d Ar flags
.Op Fl mpv
.Op Fl f Pa script
.Ar time
.Sh DESCRIPTION
The
.Nm
utility read commands from the standard input or a
specified file for execution at a later time.
.Pp
The job is executed in a separate process group by 
.Xr sh 1 ,
using the current environment (modulo some of the
obviously shell-related variables) and 
.Xr umask (1) .
.Pp
The options are as follows:
.Bl -tag -width aa-script
.It Fl d Ar flags
Turn debugging on. 
.Ar flags 
is a bitmask of values that turn on and off various settings:
.Bl -tag -width 0xff -compact
.It Ar 0x01
Print the job time.
.It Ar 0x02
Do not store that at job.
.It Ar 0x04
Print the current time before parsing the job time.
.It Ar 0x08
Print the # of seconds until the job time.
.El
.Pp
The
.Fl d
flag is intended for testing and debugging, and may produce
unexpected results if used in production.
.It Fl f Pa script
Take the commands from
.Pa script
instead of from stdin.
.It Fl m
Notify the user when the job has finished, via
.Xr write 1 
or
.Xr mail 1 ,
depending on whether the user is logged in or not.
.It Fl v
After submitting the job, display the job id and scheduled
execution time.
.El
.Sh RETURN VALUES
The
.Nm
utility exits 0 on success, and >0 if an error occurs.
.Sh TIME SPECIFICATION
.Pp
.Nm
has a fairly rich language for describing the job time.
It is roughly in the format
.Ar time Op ar date ,
where you may either specify the traditional 
.Em HH:MM MM/DD/YY
date, or more complex formats.
.Pp
.Bl -tag -width date -compact
.It Ar Time
is in one of the following forms (or blank,
which is the current time of day):
.Bl -tag -width "HH:MM [ am/pm ]"  -compact
.It Ar HH:MM Op am/pm
The given time
.It Em noon
.It Em midnight
.It Em teatime
4pm.
.El
.Pp
.It Ar Date
is in the form
.Bl -tag -width "exactly NN YY MM" -compact
.It Em MM/YY/YY Ns Op YY
Standard (US) date format
.It Em DD.MM.YY Ns Op YY
Dotted date format
.It Ar month day
.It Ar month day Ns Op Ns No , Ns Ar year 
.It Ar day Ns No - Ns Ar month Ns Op Ns No - Ns Ar year
.It Em \+ Ar count Ar offset
offset
.Ar count
 (minutes/hours/days/months/years) from 
.Ar time .
In this format, the
.Ar time
may also be the special value
.Em now
.Po
for 
.Dq now + 45 minutes ,
etc
.Pc
.It Ar count Ar offset Em from Ar day-of-week
offset 
.Ar count
(days/months/years) from
.Ar day-of-week
.Po
or the special day
.Em now ,
which means today
.Pc
.It Em today
This 
.Ar time
today.
.It Em tomorrow
This
.Ar time
tomorrow.
.It Em yesterday
This
.Ar time
yesterday
.Po only useful with the 
.Em
next
date format
.Pc
.It Em tonight
If the hour is > 18 or the hour < 6, it's that hour in the
night.
For example, 
.Dq 3am tonight
means 3am tomorrow morning.
.It Em next Ar unit
The next day/week/year.
.El
.El
.El
.Sh FILES
.Bl -tag -width @ATDIR@ -compact
.It Pa @ATDIR@
Where at jobs are spooled.
.El
.Sh EXAMPLES
.Bl -tag -width "45 weeks from tomorrow" -compact
.It Em next week
This time 7 days from now.
.It Em teatime tomorrow
4pm tomorrow.
.It Em exactly 5 weeks from now
.It Em three weeks from yesterday
.It Em teatime 12-feb-2008
.It Em noon + 3 days
.El
.Sh ERRORS
.Bl -tag -width "syntax error (...)"
.It "syntax error"
.Nm
couldn't parse the requested job time.
.It "syntax error (pluralization)"
Mismatch in units
.Po 
.Em 1 days ,
.Em 3 week ,
etc.
.Pc
.It "incorrect time of day"
The hour of the day was > 12
.Pq am/pm
or > 24.
.It "cannot travel back in time"
The specified time has already passed.
.El
.Sh SEE ALSO
.Xr atq 1 ,
.Xr atrm 1 ,
.Xr maketime 3 .
.Sh AUTHOR
.An Jessica Parsons
.Pq Li orc@pell.portland.or.us
.Sh BUGS
The time specification language may be a little byzantine.


at.c
/* Copyright 2025 by Jessica Loren Parsons (orc@pell.portland.or.us)
 * See the COPYRIGHT file included in this distribution for
 * terms of use.
 */
/*
 * at:   run a batch job at a given time (like go, but schedulable)
 */
#include "config.h"
#include <stdio.h>
#include <fcntl.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>
#include <stdarg.h>
#include <time.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <pwd.h>
#include <unistd.h>
#include <ctype.h>
#include <syslog.h>

#ifdef HAVE_LIBGEN_H
#   include <libgen.h>
#endif

#include "at.h"
#include "y.tab.h"

/* carry around the environment so the batch job can use it.
 */
extern char **environ;

static int notify = 0;		/* notify the user when the job finishes */
static int debug = 0;		/* various debugging flags for development */
static int verbose = 0;		/* print job id & job time */

static char *pgm;


#ifndef HAVE_BASENAME
/*
 * get the non-directory part of a pathname
 */
static char *
basename(char *p)
{
    char *ret = strrchr(p, '/');

    return ret ? (1+ret) : p;
}
#endif


/*
 * whine bitterly about something, then die a horrible death
 */
static void
abend(char *fmt, ...)
{
    va_list ptr;

    va_start(ptr, fmt);
    fprintf(stderr, "%s: ", pgm);
    vfprintf(stderr, fmt, ptr);
    fputc('\n', stderr);
    va_end(ptr);
    exit(1);
}


/*
 * spit out a usage message, then die
 */
static void
usage()
{
    if ( strcmp(pgm, "in") == 0 )
	fprintf(stderr, "usage: %s when << job\n", pgm);
    else
	fprintf(stderr, "usage: %s [-m] [-f file] when << job\n", pgm);
    exit(1);
}


/*
 * write a job into the at spool
 */
static void
savejob(time_t when)
{
    int c, i, fd;
    char *v, *r;
    char job[1+5+16+1];	/* 1 (id) + 8 (date) + 5 (extension) + 1 (0) */
    unsigned short seq = 0;
    FILE *output = 0 /* meaningless, but it makes gcc shut the fuck up */;
    char *pwd;
    struct passwd *user = getpwuid(getuid());
    int size;

    if ( (pwd = malloc(size=1024)) == 0 )
	abend("%s", strerror(errno));

    while ( (getcwd(pwd,size) == 0) && (errno == ERANGE) ) {
	size *= 2;
	if ( (pwd = realloc(pwd, size)) == 0 )
	    abend("%s", strerror(errno));
    }

    if ( chdir(ATDIR) != 0 )
	abend("%s", strerror(errno));

    while (1) {
	snprintf(job, sizeof job, "c%05lx%08lx", (long)seq, when/60);

	if ( (fd = open(job, O_CREAT|O_EXCL|O_WRONLY, 0600)) != -1 )
	    break;
	if ( errno != EEXIST )
	    abend("%s", strerror(errno));
	else if ( ++seq == 0 )
	    abend("spool is full -- can't save job");
    }
    if ( (fchown(fd, getuid(), getgid()) == -1)
	   || ((output = fdopen(fd, "w")) == 0) ) {
	unlink(job);
	abend("spool: %s", strerror(errno));
    }

    /* become the user before writing to the spoolfile.
     */
    setegid(getgid());
    seteuid(getuid());

    fprintf(output, "#! /bin/sh\n");
    
    /*
     * write a koeneg-at compatable header
     */
    fprintf(output, "# atrun uid=%d gid=%d\n", getuid(), getgid());
    fprintf(output, "# mail      %s %d\n", user->pw_name, notify);
    
    /* 
     * write out the environment
     */
    fprintf(output, "umask %o\n", i=umask(0777)); umask(i);
    
    for (i=0; environ[i]; i++) {
	/* weed out tty-specific variables
	 */
	if ( strncmp(environ[i], "TERM=", 5) == 0 )
	    continue;
	if ( strncmp(environ[i], "TERMCAP=", 8) == 0 )
	    continue;
	if ( strncmp(environ[i], "DISPLAY=", 8) == 0 )
	    continue;
	if ( strncmp(environ[i], "_=", 2) == 0 )
	    continue;
	/* make certain that this is a valid environment
	 * variable in the form of name=<stuff>
	 */
	for (v=environ[i]; isalnum(*v); ++v)
	    ;
	if ( *v == '=' ) {
	    /* legitimate environment variable */
	    fwrite(environ[i], v-environ[i], 1, output);
	    fputc('=',output);
	    for (r=v+1; *r; ++r) {
		if ( isprint(*r) ) {
		    if ( ! (isalnum(*r) || *r == '/' || *r == ':') )
			fputc('\\', output);
		    fputc(*r, output);
		}
		else
		    fprintf(output, "'%c'", *r);
	    }
	    fprintf(output, "; export ");
	    fwrite(environ[i], v-environ[i], 1, output);
	    fputc('\n', output);
	}
    }

    fprintf(output, "\ncd %s || exit 1\n\n", pwd);

    size = 0;
    while ( (c = getchar()) != EOF )
	if ( fputc(c, output) == EOF ) {
	    fclose(output);
	    unlink(job);
	    abend("write error");
	}
	else size ++;
    fclose(output);
    if (size == 0) {
	unlink(job);
	exit(1);
    }
    syslog(LOG_INFO, "User %d, job %s", getuid(), job);
    if (verbose)
	printf("%s %s", job, ctime(&when));
    exit(0);
}


int
main(argc, argv)
char **argv;
{
    time_t now, jobtime;
    int opt;
    int redirect = 0;

    pgm = basename(argv[0]);
    opterr = 1;
    
    if ( strcmp(pgm, "in") == 0 ) {
	argv[0] = "in";

	if ( argc < 2 )
	    usage();
    }
    else {
	while ( (opt = getopt(argc,argv, "d:f:mv")) != EOF )
	    switch (opt) {
	    case 'd':
		    debug = atoi(optarg);
		    break;
	    case 'v':
		    verbose = 1;
		    break;
	    case 'm':
		    notify = 1;
		    break;
	    case 'f':
		    if ( redirect )
			abend("too many -f options");
		    if ( freopen(optarg, "r", stdin) == 0 )
			abend("%s", strerror(errno));
	    default:
		usage();
	    }

	argc -= optind;
	argv += optind;

	if ( argc < 1 )
	    usage();
    }
       
    openlog("at", LOG_PID, LOG_CRON);
    jobtime = maketime(argc, argv, abend);

    if ( debug & (0x04|0x08) )
	time(&now);
    if ( debug & 0x04 )
	printf("now: %s", ctime(&now));
    if ( debug & 0x01 )
	printf("job: %s", ctime(&jobtime));
    if ( debug & 0x08 )
	printf("%ld\n", jobtime-now);    
    if ( !(debug & 0x02) )
	savejob(jobtime);
    exit(0);
}
